instrumentation/spring/spring-web-3.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/springweb/WebApplicationContextInstrumentation.java

```diff
@@ -59,10 +61,29 @@ public void transform(TypeTransformer transformer) {
     public static void onEnter(@Advice.Argument(0) ConfigurableListableBeanFactory beanFactory) {
       if (beanFactory instanceof BeanDefinitionRegistry
           && !beanFactory.containsBean("otelAutoDispatcherFilter")) {
+        try {
+          // Firstly check whether DispatcherServlet is present. We need to load an instrumented
+          // class from spring-webmvc to trigger injection that makes
+          // OpenTelemetryHandlerMappingFilter available.
+          beanFactory
+              .getBeanClassLoader()
+              .loadClass("org.springframework.web.servlet.DispatcherServlet");
 
-        ((BeanDefinitionRegistry) beanFactory)
-            .registerBeanDefinition(
-                "otelAutoDispatcherFilter", new HandlerMappingResourceNameFilter.BeanDefinition());
+          // Now attempt to load our injected instrumentation class.
+          Class<?> clazz =
+              beanFactory
+                  .getBeanClassLoader()
+                  .loadClass("org.springframework.web.servlet.OpenTelemetryHandlerMappingFilter");
```

Comment:
`OpenTelemetryHandlerMappingFilter` is defined in the webmvc instrumentation, right? This makes spring-web and spring-webmvc instrumentations depend on each other; one can't function without the other. 
I think it'd be more preferable to have one `InstrumentationModule` delivering a single, coherent piece of functionality (webmvc instrumentation) even if we have to depend on two libs.

Reply:
My understanding is that doesn't really work when when 2 modules are in different class loaders. Muzzle will fail on spring-web because instrumentation uses classes from spring-webmvc hence they can't be in the same `InstrumentationModule`.

Reply:
Hmm, that's true -- the instrumentation won't be applied if the classloader doesn't contain *all* references.

🤯

Reply:
do we have this problem for all instrumentations which use currently apply a muzzle `extraDependency`? is this change needed for this PR, or can we split out to discuss separately?

Reply:
If we assume that all spring jars are in the same class loader then this would not be needed. Nowadays this probably isn't super important, but years ago when spring was pushing their dm server and all spring jars had osgi manifests then when instrumenting spring it was best to make sure that the code you inject with instrumentation doesn't reference anything from bundles that instrumented bundle doesn't depend on. If we just insert reference to `OpenTelemetryHandlerMappingFilter` into something like `org.springframework.web.context.support.AbstractRefreshableWebApplicationContext` we have created a dependency from `spring-web` to `spring-webmvc` which doesn't exist in original code (dependency is the other way around) which would make application fail if it for some reason used `spring-web` but not `spring-webmvc`.
`extraDependency` is usually completely fine, as long as instrumented library has a non optional dependency to something then we can freely use it in instrumentation.

Reply:
> `extraDependency` is usually completely fine, as long as instrumented library has a non optional dependency to something then we can freely use it in instrumentation.

oh yes this makes perfect sense thx 👍👍

Reply:
just adding comment in case I come back here later 😄: we can freely _use_ the `extraDependency` in the instrumentation, but we cannot freely _instrument_ the `extraDependency`